File: packages/harness-mcp/src/tools/beads.ts
import { execCommand, findMonorepoRoot } from '../utils.js';
import type { BeadsIssue } from '../types.js';

export function getIssue(issueId: string): BeadsIssue {
  const root = findMonorepoRoot();
  // Use bd list --json with --limit 0 to get all issues
  const result = execCommand(`bd list --json --limit 0`, root);

  if (!result.success) {
    throw new Error(`Failed to list issues: ${result.output}`);
  }

  try {
    const issues = JSON.parse(result.output);
    const issue = issues.find((i: BeadsIssue) => i.id === issueId);

    if (!issue) {
      throw new Error(`Issue not found: ${issueId}`);
    }

    return issue;
  } catch (error: any) {
    if (error.message?.includes('Issue not found')) {
      throw error;
    }
    throw new Error(`Failed to parse Beads issues: ${result.output}`);
  }
}

export function listIssues(filter?: {
  status?: string;
  labels?: string[];
  priority?: string;
}): BeadsIssue[] {
  const root = findMonorepoRoot();

  let command = 'bd list --json';

  if (filter?.status) {
    command += ` --status ${filter.status}`;
  }

  if (filter?.labels?.length) {
    command += ` --label ${filter.labels.join(',')}`;
  }

  if (filter?.priority) {
    command += ` --priority ${filter.priority}`;
  }

  const result = execCommand(command, root);

  if (!result.success) {
    throw new Error(`Failed to list issues: ${result.output}`);
  }

  try {
    return JSON.parse(result.output);
  } catch (error) {
    return [];
  }
}

export function updateIssue(
  issueId: string,
  updates: {
    status?: 'pending' | 'in-progress' | 'completed' | 'blocked';
    labels?: { add?: string[]; remove?: string[] };
    notes?: string;
  }
): void {
  const root = findMonorepoRoot();

  if (updates.status) {
    const result = execCommand(`bd update ${issueId} --status ${updates.status}`, root);
    if (!result.success) {
      throw new Error(`Failed to update issue status: ${result.output}`);
    }
  }

  if (updates.labels?.add) {
    for (const label of updates.labels.add) {
      execCommand(`bd label add ${issueId} ${label}`, root);
    }
  }

  if (updates.labels?.remove) {
    for (const label of updates.labels.remove) {
      execCommand(`bd label remove ${issueId} ${label}`, root);
    }
  }

  if (updates.notes) {
    execCommand(`bd update ${issueId} --notes "${updates.notes}"`, root);
  }
}

export function closeIssue(issueId: string): void {
  const root = findMonorepoRoot();
  const result = execCommand(`bd close ${issueId}`, root);

  if (!result.success) {
    throw new Error(`Failed to close issue: ${result.output}`);
  }
}

export function getPriority(): BeadsIssue[] {
  const root = findMonorepoRoot();
  const result = execCommand('bv --robot-priority', root);

  if (!result.success) {
    throw new Error(`Failed to get priority: ${result.output}`);
  }

  try {
    return JSON.parse(result.output);
  } catch (error) {
    return [];
  }
}

File: packages/agent-sdk/src/create_something_agents/tools/beads.py
"""
Beads tool - Agent-native issue tracking.

Track work. Pick up where you left off.
Work survives session restarts, context limits, even crashes.
"""

from __future__ import annotations

import subprocess
from typing import Any, Literal

# Tool definition for Claude
beads_tool: dict[str, Any] = {
    "name": "beads",
    "description": """Track issues, dependencies, and work state with Beads.

Beads is issue tracking designed for AI agents. Work survives session restarts.

Actions:
- create: Create a new issue
- update: Update issue status
- close: Mark issue as done
- list: List all open issues
- show: Show issue details
- ready: Show unblocked work only

Labels:
- Scope: agency, io, space, ltd
- Type: feature, bug, research, refactor
- Priority: P0 (drop everything) to P4 (maybe never)

Example workflow:
1. At session start: action="ready" to see what to work on
2. When starting: action="update" with status="in-progress"
3. When done: action="close"
4. If you discover new work: action="create"
""",
    "input_schema": {
        "type": "object",
        "properties": {
            "action": {
                "type": "string",
                "enum": ["create", "update", "close", "list", "show", "ready"],
                "description": "The action to perform",
            },
            "issue_id": {
                "type": "string",
                "description": "Issue ID (required for update, close, show)",
            },
            "title": {
                "type": "string",
                "description": "Issue title (required for create)",
            },
            "priority": {
                "type": "integer",
                "enum": [0, 1, 2, 3, 4],
                "description": "Priority level: 0=drop everything, 4=maybe never",
            },
            "labels": {
                "type": "array",
                "items": {"type": "string"},
                "description": "Labels to add (scope: agency/io/space/ltd, type: feature/bug/etc)",
            },
            "status": {
                "type": "string",
                "enum": ["open", "in-progress", "code-complete", "verified"],
                "description": "Issue status (for update action)",
            },
        },
        "required": ["action"],
    },
}


def execute_beads(
    action: Literal["create", "update", "close", "list", "show", "ready"],
    issue_id: str | None = None,
    title: str | None = None,
    priority: int | None = None,
    labels: list[str] | None = None,
    status: str | None = None,
) -> str:
    """
    Execute Beads CLI commands.

    Args:
        action: The action to perform
        issue_id: Issue ID (for update, close, show)
        title: Issue title (for create)
        priority: Priority level 0-4
        labels: Labels to add
        status: Issue status (for update)

    Returns:
        Command output
    """
    try:
        if action == "create":
            if not title:
                return "[Error: title is required for create action]"

            cmd = f'bd create "{title}"'
            if priority is not None:
                cmd += f" --priority P{priority}"
            if labels:
                for label in labels:
                    cmd += f" --label {label}"

        elif action == "update":
            if not issue_id:
                return "[Error: issue_id is required for update action]"

            cmd = f"bd update {issue_id}"
            if status:
                cmd += f" --status {status}"

        elif action == "close":
            if not issue_id:
                return "[Error: issue_id is required for close action]"
            cmd = f"bd close {issue_id}"

        elif action == "list":
            cmd = "bd list --json"

        elif action == "ready":
            cmd = "bd ready --json"

        elif action == "show":
            if not issue_id:
                return "[Error: issue_id is required for show action]"
            cmd = f"bd show {issue_id}"

        else:
            return f"[Error: Unknown action: {action}]"

        # Execute command
        result = subprocess.run(
            cmd,
            shell=True,
            capture_output=True,
            text=True,
            timeout=30,
        )

        output = result.stdout
        if result.stderr:
            output += f"\n[stderr]: {result.stderr}"
        if result.returncode != 0:
            output += f"\n[exit code: {result.returncode}]"

        return output.strip() or "(no output)"

    except subprocess.TimeoutExpired:
        return "[Error: Beads command timed out]"
    except FileNotFoundError:
        return "[Error: Beads CLI (bd) not found. Is it installed?]"
    except Exception as e:
        return f"[Error: {type(e).__name__}: {e!s}]"
